packages/astro/src/core/endpoint/index.ts

```diff
@@ -91,7 +91,11 @@ export function createAPIContext({
 			if (typeof val !== 'object') {
 				throw new AstroError(AstroErrorData.LocalsNotAnObject);
 			} else {
-				Reflect.set(request, clientLocalsSymbol, val);
+				if (Reflect.has(request, 'runtime') && !Reflect.has(val, 'runtime')) {
+					throw new AstroError(AstroErrorData.LocalsNoRuntime);
+				} else {
+					Reflect.set(request, clientLocalsSymbol, val);
+				}
```

Comment:
This should be in core. Astro doesn't know anything about "runtime" and it should stay that way. We should move this code in the adapter.

Reply:
@ematipico in [Discord](https://discord.com/channels/830184174198718474/1124318874175819847/1126171117309136899) you said, we should make sure that `locals` are not overwritten by a middleware function. How would that be possible if the check is happening in the adapter?